File: iosgl/iosgl/bt/res/BaseRes.m
//
//  BaseRes.m
//  iosgl
//
//  Created by zhao on 12/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//

#import "BaseRes.h"
#import <UIKit/UIKit.h>
#import "ObjDataManager.h"
#import "ParticleManager.h"
#import "MaterialManager.h"
#import "NSData+GZIP.h"
#import "Scene_data.h"
#import <zlib.h>

@interface BaseRes()
@property(nonatomic,strong)SuccessBlock imgFun;
@property(nonatomic,assign)int imgNum;
@property(nonatomic,assign)int imgLoadNum;
@property(nonatomic,assign)BOOL imgComplete;
/*
 public static IMG_TYPE: number = 1;
 public static OBJS_TYPE: number = 2;
 public static MATERIAL_TYPE: number = 3;
 public static PARTICLE_TYPE: number = 4;
 public static SCENE_TYPE: number = 5;
 public static ZIP_OBJS_TYPE: number = 6;
 public static PREFAB_TYPE: number = 1;
 public static SCENE_PARTICLE_TYPE: number = 11;
 */
@end
@implementation BaseRes
 
typedef void (^PostSuccess)(NSDictionary *responseJson);

//解压缩
- (NSData *)gzipInflate:(NSData*)data
{
    if ([data length] == 0) return data;
    
    unsigned long full_length = [data length];
    unsigned long  half_length = [data length] / 2;
    
    NSMutableData *decompressed = [NSMutableData dataWithLength: full_length + half_length];
    BOOL done = NO;
    int status;
    
    z_stream strm;
    strm.next_in = (Bytef *)[data bytes];
    strm.avail_in = (uInt)[data length];
    strm.total_out = 0;
    strm.zalloc = Z_NULL;
    strm.zfree = Z_NULL;
    
    if (inflateInit2(&strm, (15+32)) != Z_OK)
        return nil;
    
    while (!done)
    {
        // Make sure we have enough room and reset the lengths.
        if (strm.total_out >= [decompressed length])
            [decompressed increaseLengthBy: half_length];
        strm.next_out = [decompressed mutableBytes] + strm.total_out;
        strm.avail_out = (uInt)([decompressed length] - strm.total_out);
        
        // Inflate another chunk.
        status = inflate (&strm, Z_SYNC_FLUSH);
        if (status == Z_STREAM_END)
            done = YES;
        else if (status != Z_OK)
            break;
    }
    if (inflateEnd (&strm) != Z_OK)
        return nil;
    
    // Set real length.
    if (done)
    {
        [decompressed setLength: strm.total_out];
        return [NSData dataWithData: decompressed];
    }
    else return nil;
}
-(void)read:(SuccessBlock)fun;
{
    self.imgFun=fun;
    
    int filetype = [self.byte readInt];
    NSLog(@"filetype->%d",filetype);
    NSLog(@"位置-->%d",self.byte.position);
 
    switch (filetype) {
        case IMG_TYPE:
            [self readImgs];
            break;
        case OBJS_TYPE:
            [self readObj];
            break;
        case MATERIAL_TYPE:
            [self readMaterial];
            break;
        case PARTICLE_TYPE:
            [self readParticle];
            break;
        case ZIP_OBJS_TYPE:
            [self readZipObj];
            break;
        default:
            NSLog(@"需要补充");
            
            break;
    }
}
 
-(void)countImg;
{
    self.imgLoadNum++;
    if (self.imgLoadNum == self.imgNum) {
        self.imgComplete = YES;
        [self allResCom];
    }
}
-(void)allResCom;
{
    self.imgFun(@"1");
}
-(void)read ;
{
    [self read:^(NSString* code) { }];
}
 
-(void)readObj;
{
      int objNum = [self.byte readInt];
    for(int i=0;i<objNum;i++){
 
        NSString *url =   [self.byte readUTF];
        int objsSize=  [self.byte readInt];
        NSData *objsNsdata=  [self.byte getNsDataByLen:objsSize];
        ByteArray *objsByte=  [[ByteArray alloc]init:objsNsdata];
        [[ ObjDataManager default] loadObjCom:objsByte url:url];
   
     }
    
    if (self.imgFun) {
        self.imgFun(@"1");
    }
}
-(void)readZipObj;
{
    int zipLen = [self.byte readInt];
    NSData *zipNsData=  [self.byte getNsDataByLen:zipLen];
    NSData *outputData =[self gzipInflate:zipNsData] ;
    NSLog(@"len-%d-解压后长度>%ld",zipLen,outputData.length);
    
    ByteArray *srcByte=  [[ByteArray alloc]init:outputData];
    [self readObj:srcByte];
    
}
-(void)readMaterial;
{
    int materLen = [self.byte readInt];
 
  for(int i=0;i<materLen;i++){
      NSString *materurl =   [self.byte readUTF];
      NSLog(@"materurl-->%@",materurl);
      int materSize=  [self.byte readInt];
      NSData *materNsdata=  [self.byte getNsDataByLen:materSize];
      ByteArray *materByte=  [[ByteArray alloc]init:materNsdata];
    
      [[MaterialManager default]addResByte:materurl dataByte:materByte];
      
      
  }
    
}
-(void)readParticle;
{
    int particeLen = [self.byte readInt];
    
    for(int i=0;i<particeLen;i++){
        NSString *particeurl =   [self.byte readUTF];
        NSLog(@"partice-->%@",particeurl);
        int particeSize=  [self.byte readInt];
        NSData *particeNsdata=  [self.byte getNsDataByLen:particeSize];
        ByteArray *particeByte=  [[ByteArray alloc]init:particeNsdata];
        [[ParticleManager default] addResByte:particeurl byteArray:particeByte];
    }
    
}
-(NSMutableArray*)readMaterialInfo;
{
    NSMutableArray *arr;
    int len = [self.byte readInt];
    if (len > 0) {
        arr=[[NSMutableArray alloc] init];
        for (int i = 0; i < len; i++) {
            NSMutableDictionary *mDic=[[NSMutableDictionary alloc]init];
            int type= [self.byte readInt];
            [mDic setObject:[NSString stringWithFormat:@"%d",type] forKey:@"type"];
            [mDic setObject:[self.byte readUTF] forKey:@"name"];
            switch (type) {
                case 0:
                    [mDic setObject:[self.byte readUTF] forKey:@"url"];
                    break;
                case 1:
                    [mDic setObject:  [NSString stringWithFormat:@"%f",[self.byte readFloat]]  forKey:@"x"];
                    break;
                case 2:
                    [mDic setObject:  [NSString stringWithFormat:@"%f",[self.byte readFloat]]  forKey:@"x"];
                    [mDic setObject:  [NSString stringWithFormat:@"%f",[self.byte readFloat]]  forKey:@"y"];
                    
                    break;
                case 3:
                    [mDic setObject:  [NSString stringWithFormat:@"%f",[self.byte readFloat]]  forKey:@"x"];
                    [mDic setObject:  [NSString stringWithFormat:@"%f",[self.byte readFloat]]  forKey:@"y"];
                    [mDic setObject:  [NSString stringWithFormat:@"%f",[self.byte readFloat]]  forKey:@"z"];
                    break;
                    
                default:
                    break;
            }
            [arr addObject:mDic];
        }
    }
    
    return arr;
    
}
  
-(void)readImgs;
{
    self.imgNum = [self.byte readInt];
     self.imgLoadNum = 0;
    for(int i=0;i<self.imgNum;i++){
        NSString *imgurl =   [self.byte readUTF];
        NSLog(@"imgurl-->%@",imgurl);
        int imgSize=  [self.byte readInt];
        NSLog(@"len-->%d",imgSize);
        if(imgSize&&[Scene_data default].supportBlob){
              NSData *imgNsdata=  [self.byte getNsDataByLen:imgSize];
        }
 
      
        
        [self countImg];
    }
}
-(void)readObj:(ByteArray *)srcByte;
{
    int objlen = [srcByte readInt];
    for(int i=0;i<objlen;i++){
        NSString *objurl =   [srcByte readUTF];
        NSLog(@"objurl-->%@",objurl);
        int objSize=  [srcByte readInt];
        NSData *objNsdata=  [srcByte getNsDataByLen:objSize];
        ByteArray *objByte=  [[ByteArray alloc]init:objNsdata];
        [[ObjDataManager default] loadObjCom:objByte url:objurl];
        
        
    }
    NSLog(@"----------" );
}
+(NSMutableArray*)readIntForTwoByte:(ByteArray*)srcByte nsdata:(NSMutableData*)nsdata ;
{
      NSMutableArray *idxItem=[[NSMutableArray alloc]init];
    int iLen = [srcByte readInt];
    int tempIdx;
    for (int i = 0; i < iLen; i++) {
        tempIdx=  [srcByte readShort];
        [idxItem addObject:[NSNumber numberWithInt:tempIdx]];
    }
    return idxItem;
}
+(NSMutableArray*)readBytes2ArrayBuffer:(ByteArray*)srcByte nsdata:(NSMutableData*)nsdata  dataWidth:(int)dataWidth   offset:(int)offset   stride:(int)stride   readType:(int)readType  ;
{
    NSMutableArray *vItem=[[NSMutableArray alloc]init];
    int verLength = [srcByte readInt];
    if (verLength <= 0) {
        return nil;
    }
    float scaleNum=1.0;
    if (readType == 0) {
        scaleNum = [srcByte readFloat];
    }
    float tempNum;
    int readNum = verLength / dataWidth;
    for (int i = 0; i < readNum; i++) {
        int pos = stride * i + offset*4;
        for (int j = 0; j < dataWidth; j++) {
            switch (readType) {
                case 0:
                    tempNum=  [srcByte readFloatTwoByte:scaleNum];
                    [vItem addObject:[NSNumber numberWithFloat:tempNum]];
                    [nsdata replaceBytesInRange:NSMakeRange(pos+j*4, 4) withBytes: &tempNum length:4];
                    break;
                case 1:
                    tempNum=  [srcByte readFloatOneByte];
                     [vItem addObject:[NSNumber numberWithFloat:tempNum]];
                    break;
                case 2:
                    tempNum=  [srcByte readByte];
                     [vItem addObject:[NSNumber numberWithFloat:tempNum]];
                    break;
                case 4:
                    tempNum=  [srcByte readFloat];
                    [vItem addObject:[NSNumber numberWithFloat:tempNum]];
                    [nsdata replaceBytesInRange:NSMakeRange(pos+j*4, 4) withBytes: &tempNum length:4];
                    break;
                default:
                    NSLog(@"没有没有");
                    break;
            }
        }
    }
    
    return vItem;
 
    
}
  //读取材质参数
+(NSArray<NSDictionary*>*)readMaterialParamData:(ByteArray*)byte;
{
    int mpNum=[byte readInt];
    if (mpNum > 0) {
           NSMutableArray<NSMutableDictionary*>* mpAry  =[[NSMutableArray alloc]init];
           for (int j = 0; j < mpNum; j++) {
               NSMutableDictionary* obj =[[NSMutableDictionary alloc]init];
               obj[@"name"] = [byte readUTF];
               obj[@"type"] = [NSNumber numberWithInt:[byte readByte]];
               if ([obj[@"type"]intValue] == 0) {
                   obj[@"url"] = [byte readUTF];
               } else if ([obj[@"type"] intValue]== 1) {
                   obj[@"x"]  =  [NSNumber numberWithInt:[byte readFloat]];
               } else if ([obj[@"type"]intValue] == 2) {
                   obj[@"x"] =  [NSNumber numberWithInt:[byte readFloat]];
                   obj[@"y"] =  [NSNumber numberWithInt:[byte readFloat]];
               } else if ([obj[@"type"]intValue] == 3) {
                   obj[@"x"]=  [NSNumber numberWithInt:[byte readFloat]];
                   obj[@"y"]=  [NSNumber numberWithInt:[byte readFloat]];
                   obj[@"z"] =  [NSNumber numberWithInt:[byte readFloat]];
               }
               [mpAry addObject:obj];
           }
        return mpAry;
    }
     
    return nil;
}
/*
 //读取材质参数
 public static readMaterialParamData(byte: ByteArray): Array<any> {
     var mpNum: number = byte.readInt();
     if (mpNum > 0) {
         var mpAry: Array<any> = new Array;
         for (var j: number = 0; j < mpNum; j++) {
             var obj: any = new Object;
             obj.name = byte.readUTF();
             obj.type = byte.readByte();
             if (obj.type == 0) {
                 obj.url = byte.readUTF();
             } else if (obj.type == 1) {
                 obj.x = byte.readFloat();
             } else if (obj.type == 2) {
                 obj.x = byte.readFloat();
                 obj.y = byte.readFloat();
             } else if (obj.type == 3) {
                 obj.x = byte.readFloat();
                 obj.y = byte.readFloat();
                 obj.z = byte.readFloat();
             }
             mpAry.push(obj);
         }
         return mpAry
     }
     return null
 }
 */
@end

File: test/four/app/src/main/java/com/z3d/res/BaseRes.java
package com.z3d.res;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.z3d.base.ByteArray;
import com.z3d.base.CallBackFun;
import com.z3d.base.Scene_data;
import com.z3d.engine.ResCount;
import com.z3d.scene.Scene3D;
import com.z3d.units.TimeUtil;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.Inflater;

public class BaseRes extends ResCount {
    public static int IMG_TYPE = 1;
    public static int OBJS_TYPE = 2;
    public static int MATERIAL_TYPE = 3;
    public static int PARTICLE_TYPE = 4;
    public static int SCENE_TYPE = 5;
    public static int ZIP_OBJS_TYPE = 6;
    public static int PREFAB_TYPE = 1;
    public static int SCENE_PARTICLE_TYPE = 11;
    protected ByteArray _byte;
    protected int imgNum;
    protected int imgLoadNum;
    private CallBackFun imgFun;

    public int version;

    public  BaseRes(Scene3D val  ){
        super(val);

    }
    public void read( ){

        this.read(null);
    }
    public void read(CallBackFun bfun){

        this.imgFun=bfun;
        int fileType   = this._byte.readInt();
        Log.d("fileType",   fileType+"" );
        if (fileType == BaseRes.IMG_TYPE) {
            this.readImg();
        } else if (fileType == BaseRes.OBJS_TYPE) {
            this.readObj(this._byte);
        } else if (fileType == BaseRes.MATERIAL_TYPE) {
            this.readMaterial();
        } else if (fileType == BaseRes.PARTICLE_TYPE) {
            this.readParticle();
        } else if (fileType == BaseRes.ZIP_OBJS_TYPE) {
            this.readZipObj();
        }else
        {
            Log.d("fileType",   "需要补充" );
        }
        if(this.imgFun!=null){
            this.imgFun.StateChange(true);
        }
    }
    private  void readMaterial()
    {
        int objNum=this._byte.readInt();
        float time = TimeUtil.getTimer();
        for (int i = 0; i < objNum; i++) {

            String url=this._byte.readUTF();
            int materialSize=this._byte.readInt();
            byte[] materialByte= this._byte.readBytes(materialSize);

            Log.d("material地址 ->",   url+" " );
            Log.d("material大小 ->",   materialSize+" " );

            scene3D.materialManager.addResByte(url, new ByteArray(materialByte));

        }

    }

    private  void readParticle()
    {
        int particleNum=this._byte.readInt();
        float time = TimeUtil.getTimer();
        for (int i = 0; i < particleNum; i++) {

            String url=this._byte.readUTF();
            int particSize=this._byte.readInt();
            byte[] particByte= this._byte.readBytes(particSize);
            Log.d("partic地址 ->",   url+" " );
            Log.d("partic大小 ->",   particSize+" " );

            scene3D.particleManager.addResByte(url, new ByteArray(particByte));

        }

    }

    //读取材质参数
    public  static  List<MaterialInfoVo> readMaterialParamData(ByteArray _byte)
    {
        int mpNum = _byte.readInt();
        if (mpNum > 0) {


            List<MaterialInfoVo> mpAry  = new ArrayList<>();
            for (int j = 0; j < mpNum; j++) {
                MaterialInfoVo obj = new MaterialInfoVo();
//                obj.put("name",_byte.readUTF());
                obj.name=_byte.readUTF();
                int  objType=_byte.readByte();
                if (objType == 0) {
//                    obj.put("type",0);
//                    obj.put("url", _byte.readUTF());
                    obj.type=0;
                    obj.url=_byte.readUTF();
                } else if (objType== 1) {
//                    obj.put("type",1);
//                    obj.put("x",_byte.readFloat());
                    obj.type=1;
                    obj.x=_byte.readFloat();
                } else if (objType== 2) {
//                    obj.put("type",2);
//                    obj.put("x", _byte.readFloat());
//                    obj.put("y",_byte.readFloat());
                    obj.type=2;
                    obj.x=_byte.readFloat();
                    obj.y=_byte.readFloat();
                } else if (objType == 3) {
//                    obj.put("type",3);
//                    obj.put("x", _byte.readFloat());
//                    obj.put("y", _byte.readFloat());
//                    obj.put("z",_byte.readFloat());
                    obj.type=3;
                    obj.x=_byte.readFloat();
                    obj.y=_byte.readFloat();
                    obj.z=_byte.readFloat();
                }
                mpAry.add(obj);
            }
            return mpAry;
        }
        return  null;
    }

    public List<MaterialInfoVo> readMaterialInfo()
    {

        int len = this._byte.readInt();
        if (len > 0) {
            List<MaterialInfoVo> $arr  = new ArrayList();
            for (int i = 0; i < len; i++) {
                MaterialInfoVo $temp  = new MaterialInfoVo();
                $temp.type = this._byte.readInt();
                $temp.name = this._byte.readUTF();
                if ($temp.type == 0) {
                    $temp.url = this._byte.readUTF();
                }
                if ($temp.type == 1) {
                    $temp.x = this._byte.readFloat();
                }
                if ($temp.type == 2) {
                    $temp.x = this._byte.readFloat();
                    $temp.y = this._byte.readFloat();
                }
                if ($temp.type == 3) {
                    $temp.x = this._byte.readFloat();
                    $temp.y = this._byte.readFloat();
                    $temp.z = this._byte.readFloat();
                }

                $arr.add($temp);
            }
            return $arr;
        } else {
            return null;
        }
    }
    public ByteArray getZipByte(ByteArray _byte)
    {
        int zipLen = this._byte.readInt();
        byte[] zipByte= this._byte.readBytes(zipLen);
        byte[] outByte=this.getZipData(zipByte);
        return  new ByteArray(outByte);
    }
    private  void   readZipObj(){
        int zipLen = this._byte.readInt();
        byte[] zipByte= this._byte.readBytes(zipLen);
        byte[] outByte=this.getZipData(zipByte);
        this.readObj(new ByteArray(outByte));
    }
    public byte[] getZipData(byte[] sourcesByte) {
        int dataLength = 1024*1024;
        byte[] result=new byte[dataLength];
        try {
            Inflater decompresser=new Inflater();
            decompresser.setInput(sourcesByte,0,sourcesByte.length);
            int resultLength = decompresser.inflate(result);
            decompresser.end();
        } catch(Exception ex) {}
        return result;

    }

    private  void  readObj(ByteArray srcByte){
        int objNum  = srcByte.readInt();
        for (int i   = 0; i < objNum; i++) {
            int pos=srcByte.byteBuffer.position;
            String url = srcByte.readUTF();
            int objLen = srcByte.readInt();
            byte[] objByte= srcByte.readBytes(objLen);
            Log.d("obj地址 ->",   url+" " );
            Log.d("obj大小 ->",   objLen+" " );

            scene3D.objDataManager.loadObjCom(  objByte, url);

        }



    }

    public static void readIntForTwoByte(ByteArray  $byte, List indexs)
    {
        int iLen   = $byte.readInt();
        for (int i = 0; i < iLen; i++) {
            indexs.add($byte.readShort());
        }
    }

    public static List<Float>  readBytes2ArrayBuffer(ByteArray $byte, ByteBuffer $data,int dataWidth, int offset, int stride, int readType)
    {
        int verLength = $byte.readInt();

        List<Float>  list=new ArrayList();

        if (verLength <= 0) {
            return list;
        }

        float scaleNum=0;
        if (readType == 0) {
            scaleNum = $byte.readFloat();
        }
        float tempNum=0;
        int readNum   = verLength / dataWidth;

        for (int i = 0; i < readNum; i++) {
            int pos = stride * i + offset;
            for (int j = 0; j < dataWidth; j++) {
                if (readType == 0) {
                    tempNum=  $byte.readFloatTwoByte(scaleNum);
                    // $data.setFloat32((pos + j) * 4, $byte.readFloatTwoByte(scaleNum), true);
                } else if (readType == 1) {
                    // $data.setFloat32((pos + j) * 4, $byte.readFloatOneByte(), true);
                    tempNum=  $byte.readFloatOneByte();
                } else if (readType == 2) {
                    // $data.setFloat32((pos + j) * 4, $byte.readByte(), true);
                    tempNum=  $byte.readByte();
                } else if (readType == 3) {
                    // $data.setFloat32((pos + j) * 4, ($byte.readByte() + 128) / 255, true);

                    tempNum=  $byte.readByte()+128.0f/255.0f;
                } else if (readType == 4) {
                    // $data.setFloat32((pos + j) * 4, $byte.readFloat(), true);

                    tempNum=  $byte.readFloat();
                }

                list.add(tempNum);

            }


        }
        return list;


    }
    public  void readImg()
    {
        this.imgNum = this._byte.readInt();
        this._byte.tracePostion("图片数量"+imgNum);
        this.imgLoadNum = 0;
        for (int i = 0; i < this.imgNum; i++) {
            String url   =  this._byte.readUTF();
            this._byte.tracePostion("b");
            int imgSize = this._byte.readInt();
            Log.d("图片地址 ->",   url+" " );
            Log.d("图片大小 ->",   imgSize+" " );
            if(imgSize>0){
                byte[] imgByte=  this._byte.readBytes(imgSize);
                Bitmap bitmap = BitmapFactory.decodeByteArray(imgByte, 0, imgByte.length);
                scene3D.textureManager.addRes(Scene_data.fileRoot+ url,bitmap);
            }
        }
        Log.d("url ->",     "-----" );
    }




}
